src/devices/src/pseudo/init.c

```diff
@@ -0,0 +1,40 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <sys/types.h>
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <unistd.h>
+
+#ifdef __x86_64__
+#define MAGIC_MMIO_SIGNAL_GUEST_BOOT_COMPLETE 0xd0000000
+#endif
+#ifdef __aarch64__
+#define MAGIC_MMIO_SIGNAL_GUEST_BOOT_COMPLETE 0x40001000
+#endif
+
+#define MAGIC_VALUE_SIGNAL_GUEST_BOOT_COMPLETE 123
+
+int main () {
+
+   int fd = open("/dev/mem", (O_RDWR | O_SYNC));
+   int mapped_size = getpagesize();
+
+   char *map_base = mmap(NULL,
+        mapped_size,
+        PROT_WRITE,
+        MAP_SHARED,
+        fd,
+        MAGIC_MMIO_SIGNAL_GUEST_BOOT_COMPLETE);
+
+   char *virt_addr = (char*)map_base;
+
+   *(volatile char*)virt_addr = MAGIC_VALUE_SIGNAL_GUEST_BOOT_COMPLETE;
```

Comment:
Marking this as `volatile` isn't necessary, and neither does it guarantee that the write will go through without IO buffering.

I found this thread to be useful: https://stackoverflow.com/questions/45753903/memory-mapped-files-and-pointers-to-volatile-objects

In this case you could do an `msync()` after the assignment to flush the data to `/dev/mem`:

```suggestion
   *map_base = MAGIC_VALUE_SIGNAL_GUEST_BOOT_COMPLETE;
   msync(map_base, mapped_size, MS_ASYNC);
```

You can use either `MS_ASYNC` or `MS_SYNC` here. I went for `MS_ASYNC` since we don't **need** to wait for the flush before continuing our init process.

Reply:
Thanks!!  Sorry, it was a bit cut and pasty.  The link helped :-)